libraries/chain/controller.cpp

```diff
@@ -4316,8 +4313,76 @@ struct controller_impl {
       }
    }
 
+   // This is only used during Savanna transition, which is a one-time occurrence,
+   // and it is only used by SHiP..
+   // It is OK to calculate from Savanna Genesis block for each Transition block.
+   std::optional<finality_data_t> get_transition_block_finality_data(const block_state_legacy_ptr& head) const {
+      fork_database_legacy_t::branch_t legacy_branch;
+      block_state_legacy_ptr legacy_root;
+      fork_db.apply_l<void>([&](const auto& forkdb) {
+         legacy_root = forkdb.root();
+         legacy_branch = forkdb.fetch_branch(head->id());
+      });
+
+      block_state_ptr prev;
+      auto bitr = legacy_branch.rbegin();
+
+      // get_transition_block_finality_data is called by SHiP as a result
+      // of receiving accepted_block signal. That is before
+      // the call to log_irreversible where root() is updated.
+      // Search both root and legacy_branch for the first block having
+      // instant_finality_extension -- the Savanna Genesis Block.
+      // Then start from the Savanna Genesis Block to create corresponding
+      // Savanna blocks.
+      if (legacy_root->header.contains_header_extension(instant_finality_extension::extension_id())) {
+         prev = block_state::create_if_genesis_block(*legacy_root);
+      } else {
+         for (; bitr != legacy_branch.rend(); ++bitr) {
+            if ((*bitr)->header.contains_header_extension(instant_finality_extension::extension_id())) {
+               prev = block_state::create_if_genesis_block(*(*bitr));
+               ++bitr;
+               break;
+            }
+         }
+      }
+
+      assert(prev);
+      const bool skip_validate_signee = true; // validated already
+
+      for (; bitr != legacy_branch.rend(); ++bitr) {
+         auto new_bsp = std::make_shared<block_state>(
+               *prev,
+               (*bitr)->block,
+               protocol_features.get_protocol_feature_set(),
+               validator_t{}, skip_validate_signee);
+
+         assert((*bitr)->action_mroot_savanna);
+         new_bsp->action_mroot = *((*bitr)->action_mroot_savanna); // required by finality_data
```

Comment:
I see we set `new_bsp->action_mroot` here, but I don't see where it is used when we create the new `block_state` at the next loop iteration?

Reply:
`action_mroot` is not used during this process of transitions; it is used when we do `prev->get_finality_data();` at line 4366.

Reply:
So if we use only the one from the last block_state, why do we populate (and did `calculate_merkle`) for all the preceding blocks?

Reply:
Oha, we could have checked if it is the last block in legacy_branch, save it, and do the copy after the for loop. I think we should still cache action_mroot in block_state_legacy because it is used several places.

Reply:
Wait, I'm totally confused. Should,'t the `action_mroot` be used in `block_state::new_valid` to compute the new `valid_t` stored in the block state, which is then used for the next `valid` computation.
I thought we had to go through the bsp from the genesis block to recreate this. If we don't why do we need the loop at line 4352?

Reply:
For finality_data, we don't need `valid`. We only need information from header state for `base_digest` plus new action_mroot. That's why `valid` is not re-created.

Reply:
At line 4360, we set `new_bsp->action_mroot = *((*bitr)->action_mroot_savanna);`.
``new_bsp->action_mroot` is not the result of `calculate_merkle(action_receipt_digests)`, it should be `parent.get_finality_mroot_claim(qc_data.qc_claim);` I think.

Reply:
No, for finality_data sent to SHiP, the action_mroot is the mroot of the actual actions, not the action_mroot in header state changed to represent qc_claim. The actual mroot is defined in line 81 of block_state.hpp

Reply:
OK, thanks!